internals/cli/masker/writer.go

```diff
@@ -155,49 +166,48 @@ func (mw *MaskedWriter) flushBuffer() {
 //
 // This should be run in a separate goroutine.
 func (mw *MaskedWriter) Run() {
+	go mw.write()
 	masking := false
 	for {
 		select {
-		case <-time.After(mw.timeout):
-			mw.lock.Lock()
-			if len(mw.output) == 0 {
-				for _, matcher := range mw.matchers {
-					matcher.Reset()
-				}
-				mw.flushBuffer()
-			}
-			mw.lock.Unlock()
 		case output := <-mw.output:
 			for _, b := range output {
 				var err error
 				if b.masked {
 					if !masking {
 						_, err = mw.w.Write([]byte(mw.maskString))
 						if err != nil {
-							mw.err = err
+							mw.err <- err
 							return
 						}
 					}
 					masking = true
 				} else {
 					_, err = mw.w.Write([]byte{b.byte})
 					if err != nil {
-						mw.err = err
+						mw.err <- err
 						return
 					}
 					masking = false
 				}
 			}
-			mw.nOut += int64(len(output))
+			mw.wg.Add(-len(output))
+		case <-time.After(mw.timeout):
+			// force the buffer to flush if not already done so.
+			select {
+			case mw.forceFlushBuffer <- struct{}{}:
+			default:
+			}
 		}
 	}
 }
 
 // Flush is called to make sure that all output is written to the underlying io.Writer.
 // Returns any errors caused by the writing.
```

Comment:
Does it return any errors or does it return the first error that occurred during writing? 

```go
// Flush ensures that all output is written to the underlying io.Writer 
// and returns an error when a MaskedWriter.Write call fails.
```

Also, what is the effect of calling flush on the writer? Does it then no longer accept Write calls or will it keep working? If so, let's be clear about that in the comment. 

Finally, just by looking at the code, I'd expect the `Flush` function to call e.g. `forceFlush` but instead it just waits until the `MaskedWriter` is done processing. It actually doesn't flush a buffer to the output, that is already done by the run logic itself. 

Take a look at e.g. the [tabwriter](https://golang.org/pkg/text/tabwriter/). That too implements an `io.Writer` and has a `Flush` function. Instead of continuously passing on the data, it buffers it and adds tabs where necessary until `Flush` is called and the buffer is written to the underlying writer. With that architecture in mind, how do you see the architecture of the `MaskedWriter` and could we want to change that? E.g. should we consider making the `MaskedWriter` buffered with a real `Flush` function and separating the `Run` logic from it in e.g. a helper function or type? Something like `masker.Stream` with functions `Stream.Start` (maps to `MaskedWriter.Run`) and `Stream.Stop`(maps to `MaskedWriter.Flush`)?